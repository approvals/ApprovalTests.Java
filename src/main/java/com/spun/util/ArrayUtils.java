package com.spun.util;

import java.lang.reflect.Array;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Vector;

import org.lambda.functions.Function1;

/**
 * A static class of convenience methods for arrays and collections.
 **/
public class ArrayUtils
{
  /************************************************************************/
  public static <T> java.util.Collection<T> addArray(java.util.Collection<T> v, T array[])
  {
    if ((array == null) || (v == null)) { return v; }
    for (int i = 0; i < array.length; i++)
    {
      v.add(array[i]);
    }
    return v;
  }
  public static <T> String toString(T[] values, Function1<T, String> formatter)
  {
    return toString(Arrays.asList(values), formatter);
  }
  public static <T> String toString(Iterable<T> values, Function1<T, String> formatter)
  {
    StringBuffer b = new StringBuffer();
    for (T t : values)
    {
      b.append(formatter.call(t) + "\r\n");
    }
    return b.toString();
  }
  /***********************************************************************/
  public static Vector toReverseVector(Vector<Object> vector)
  {
    Vector<Object> reverse = new Vector<Object>(vector.size());
    for (int i = vector.size() - 1; i >= 0; i--)
    {
      reverse.add(vector.elementAt(i));
    }
    return reverse;
  }
  /***********************************************************************/
  public static <T> T[] toReverseArray(T[] array)
  {
    for (int i = 0; i < array.length / 2; i++)
    {
      T o1 = array[i];
      int end = array.length - i - 1;
      T o2 = array[end];
      array[i] = o2;
      array[end] = o1;
    }
    return array;
  }
  /***********************************************************************/
  public static <T> T[] addToArray(T[] array, T object)
  {
    Object[] newArray = null;
    if (array == null)
    {
      newArray = (Object[]) Array.newInstance(object.getClass(), 1);
      newArray[0] = object;
    }
    else
    {
      newArray = (Object[]) Array.newInstance(array.getClass().getComponentType(), array.length + 1);
      System.arraycopy(array, 0, newArray, 0, array.length);
      newArray[array.length] = object;
    }
    return (T[]) newArray;
  }
  /***********************************************************************/
  public static <T> T[] getSubsection(T[] array, int startInclusive, int endExclusive)
  {
    int length = endExclusive - startInclusive;
    length = length > array.length ? array.length : length;
    T[] newArray = (T[]) Array.newInstance(array.getClass().getComponentType(), length);
    System.arraycopy(array, startInclusive, newArray, 0, length);
    return newArray;
  }
  /***********************************************************************/
  public static boolean isEmpty(Object[] array)
  {
    return ((array == null) || (array.length == 0));
  }
  /************************************************************************/
  public static boolean isEmpty(Collection collection)
  {
    return ((collection == null) || (collection.size() == 0));
  }
  /***********************************************************************/
  public static <T> T getSingleton(T[] parts)
  {
    if (parts == null) { return null; }
    switch (parts.length)
    {
      case 0 :
        return null;
      case 1 :
        return parts[0];
      default :
        throw new Error("Called with more than one object in the array " + Arrays.asList(parts));
    }
  }
  /***********************************************************************/
  public static <H, T extends H> T getFirst(T[] array, Comparator<H> compartor)
  {
    return get(array, compartor, true);
  }
  /***********************************************************************/
  public static <H, T extends H> T getFirst(Collection<T> array, Comparator<H> sorter)
  {
    return get((T[]) array.toArray(), sorter, true);
  }
  /***********************************************************************/
  public static <T> T getLast(T[] array, Comparator<T> sorter)
  {
    return get(array, sorter, false);
  }
  /***********************************************************************/
  public static <T> T getLast(Collection<T> array, Comparator<T> sorter)
  {
    return get((T[]) array.toArray(), sorter, false);
  }
  /***********************************************************************/
  private static <H, T extends H> T get(T[] array, Comparator<H> sorter, boolean wantFirst)
  {
    if (isEmpty(array)) { return null; }
    T last = array[0];
    for (int i = 1; i < array.length; i++)
    {
      int compare = sorter.compare(last, array[i]);
      if ((wantFirst && compare > 0) || (!wantFirst && compare < 0))
      {
        last = array[i];
      }
    }
    return last;
  }
  /***********************************************************************/
  public static List combineResults(Object[] array, String invokeMethod)
  {
    if (ArrayUtils.isEmpty(array)) { return Collections.EMPTY_LIST; }
    try
    {
      return combineResults(array, ObjectUtils.getGreatestCommonDenominator(array, invokeMethod));
    }
    catch (Exception e)
    {
      throw ObjectUtils.throwAsError(e);
    }
  }
  /***********************************************************************/
  @SuppressWarnings("rawtypes")
  public static List combineResults(Object[] array, Method method)
  {
    if (isEmpty(array)) { return new ArrayList(0); }
    try
    {
      ArrayList<Object> list = new ArrayList<Object>();
      for (int i = 0; i < array.length; i++)
      {
        addArray(list, (Object[]) method.invoke(array[i], (Object[]) null));
      }
      return list;
    }
    catch (Throwable e)
    {
      throw ObjectUtils.throwAsError(e);
    }
  }
  /***********************************************************************/
  public static Object[] combine(Object[] a, Object[] b)
  {
    if (isEmpty(a) && isEmpty(b))
    {
      if (a != null)
      {
        return (Object[]) Array.newInstance(a.getClass().getComponentType(), 0);
      }
      else if (b != null)
      {
        return (Object[]) Array.newInstance(b.getClass().getComponentType(), 0);
      }
      else
      {
        return new Object[0];
      }
    }
    else if (isEmpty(a))
    {
      return b;
    }
    else if (isEmpty(b))
    {
      return a;
    }
    else
    {
      Object[] newArray = null;
      newArray = (Object[]) Array.newInstance(a.getClass().getComponentType(), a.length + b.length);
      System.arraycopy(a, 0, newArray, 0, a.length);
      System.arraycopy(b, 0, newArray, a.length, b.length);
      return newArray;
    }
  }
  /***********************************************************************/
  public static <T> boolean contains(T[] values, T value)
  {
    for (int i = 0; i < values.length; i++)
    {
      if (value.equals(values[i])) { return true; }
    }
    return false;
  }
  /***********************************************************************/
  public static boolean contains(int[] values, int value)
  {
    for (int i = 0; i < values.length; i++)
    {
      if (value == values[i]) { return true; }
    }
    return false;
  }
  /************************************************************************/
  public static <T> T getLast(List<T> list)
  {
    return list.get(list.size() - 1);
  }
  /************************************************************************/
  public static <K, T> T getDefault(HashMap<K, T> map, K key, T defaultValue)
  {
    T value = map.get(key);
    if (value == null)
    {
      map.put(key, defaultValue);
      value = defaultValue;
    }
    return value;
  }
  /************************************************************************/
  public static <K, V> int countValues(HashMap<K, V> out, V matching)
  {
    return count(matching, out.values());
  }
  /************************************************************************/
  public static <V> int count(V matching, Collection<V> values)
  {
    int count = 0;
    for (V value : values)
    {
      if (matching.equals(value))
      {
        count++;
      }
    }
    return count;
  }
  public static <T> List<T> combine(List<T> list1, List<T> list2)
  {
    List<T> all = new ArrayList<T>();
    all.addAll(list1);
    all.addAll(list2);
    return all;
  }
  public static <T> Iterable<T> asIterable(Iterator<T> iterator)
  {
    return new IterableWrapper(iterator);
  }
  public static class IterableWrapper<T> implements Iterable<T>
  {
    private final Iterator<T> iterator;
    public IterableWrapper(Iterator<T> iterator)
    {
      this.iterator = iterator;
    }
    @Override
    public Iterator<T> iterator()
    {
      return iterator;
    }
  }
}