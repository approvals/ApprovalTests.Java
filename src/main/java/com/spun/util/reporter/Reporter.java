package com.spun.util.reporter;

/***********************************************************************/
public interface Reporter
{

  /***********************************************************************/
  
  void throwError(Throwable error);
}
