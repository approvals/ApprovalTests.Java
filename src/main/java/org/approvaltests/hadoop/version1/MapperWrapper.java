package org.approvaltests.hadoop.version1;

import java.io.IOException;

import org.apache.hadoop.mapreduce.Mapper;

public class MapperWrapper<KeyIn, ValueIn, KeyOut, ValueOut> extends SmartMapper<KeyIn, ValueIn, KeyOut, ValueOut>
{
  private final Class<KeyIn>                             keyIn;
  private final Class<ValueIn>                           valueIn;
  private final Class<KeyOut>                            keyOut;
  private final Class<ValueOut>                          valueOut;
  private final Mapper<KeyIn, ValueIn, KeyOut, ValueOut> mapper;
  public MapperWrapper(Mapper<KeyIn, ValueIn, KeyOut, ValueOut> mapper, Class<KeyIn> keyIn,
      Class<ValueIn> valueIn, Class<KeyOut> keyOut, Class<ValueOut> valueOut)
  {
    this.mapper = mapper;
    this.keyIn = keyIn;
    this.valueIn = valueIn;
    this.keyOut = keyOut;
    this.valueOut = valueOut;
  }
  @Override
  public Class<KeyIn> getKeyInType()
  {
    return keyIn;
  }
  @Override
  public Class<ValueIn> getValueInType()
  {
    return valueIn;
  }
  @Override
  public Class<KeyOut> getKeyOutType()
  {
    return keyOut;
  }
  @Override
  public Class<ValueOut> getValueOutType()
  {
    return valueOut;
  }
  @Override
  public void run(org.apache.hadoop.mapreduce.Mapper.Context context) throws IOException, InterruptedException
  {
    mapper.run(context);
  }
}
