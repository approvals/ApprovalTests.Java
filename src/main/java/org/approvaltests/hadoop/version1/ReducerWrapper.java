package org.approvaltests.hadoop.version1;

import java.io.IOException;

import org.apache.hadoop.mapreduce.Reducer;

public class ReducerWrapper<KeyIn, ValueIn, KeyOut, ValueOut>
    extends
      SmartReducer<KeyIn, ValueIn, KeyOut, ValueOut>
{
  private final Class<KeyIn>                              keyIn;
  private final Class<ValueIn>                            valueIn;
  private final Class<KeyOut>                             keyOut;
  private final Class<ValueOut>                           valueOut;
  private final Reducer<KeyIn, ValueIn, KeyOut, ValueOut> reducer;
  public ReducerWrapper(Reducer<KeyIn, ValueIn, KeyOut, ValueOut> reducer, Class<KeyIn> keyIn,
      Class<ValueIn> valueIn, Class<KeyOut> keyOut, Class<ValueOut> valueOut)
  {
    this.reducer = reducer;
    this.keyIn = keyIn;
    this.valueIn = valueIn;
    this.keyOut = keyOut;
    this.valueOut = valueOut;
  }
  @Override
  public Class<KeyIn> getKeyInType()
  {
    return keyIn;
  }
  @Override
  public Class<ValueIn> getValueInType()
  {
    return valueIn;
  }
  @Override
  public Class<KeyOut> getKeyOutType()
  {
    return keyOut;
  }
  @Override
  public Class<ValueOut> getValueOutType()
  {
    return valueOut;
  }
  @Override
  public void run(Context context) throws IOException, InterruptedException
  {
    reducer.run(context);
  }
}
