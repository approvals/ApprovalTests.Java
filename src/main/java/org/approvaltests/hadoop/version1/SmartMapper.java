package org.approvaltests.hadoop.version1;

import org.apache.hadoop.mapreduce.Mapper;
import org.approvaltests.hadoop.Transform;

public abstract class SmartMapper<KeyIn, ValueIn, KeyOut, ValueOut>
    extends
      Mapper<KeyIn, ValueIn, KeyOut, ValueOut> implements Transform<KeyIn, ValueIn, KeyOut, ValueOut>
{
}
