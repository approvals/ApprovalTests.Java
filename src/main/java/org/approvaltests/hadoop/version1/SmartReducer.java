package org.approvaltests.hadoop.version1;

import org.apache.hadoop.mapreduce.Reducer;
import org.approvaltests.hadoop.Transform;

public abstract class SmartReducer<KeyIn, ValueIn, KeyOut, ValueOut>
    extends
      Reducer<KeyIn, ValueIn, KeyOut, ValueOut> implements Transform<KeyIn, ValueIn, KeyOut, ValueOut>
{
}