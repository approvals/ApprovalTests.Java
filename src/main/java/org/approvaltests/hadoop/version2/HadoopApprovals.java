package org.approvaltests.hadoop.version2;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mrunit.MapDriver;
import org.apache.hadoop.mrunit.MapReduceDriver;
import org.apache.hadoop.mrunit.ReduceDriver;
import org.approvaltests.Approvals;
import org.approvaltests.hadoop.Echo;
import org.approvaltests.hadoop.PairComparer;
import org.approvaltests.hadoop.WritableUtils;

import com.spun.util.ArrayUtils;

/**
 * To use this package you will need:
 * <ul>
 * <li>hadoop-client-1.0.3.jar</li>
 * <li>hadoop-core-0.20.2.jar</li>
 * <li>hadoop-mapred-test-0.22.0.jar</li>
 * <li>log4j-1.2.15.jar</li>
 * <li>mockito-all-1.8.5.jar</li>
 * <li>mrunit-0.9.0-incubating-hadoop1.jar</li>
 * </ul>
 * in your build path.
 * 
 * You will also want to either extended the SmartMapper & SmartWrapper classes,or use the MapperWrapper & ReducerWrapper classes.
 * 
 * @see SmartMapper
 * @see SmartReducer
 * @see MapperWrapper
 * @see ReducerWrapper
 */
@SuppressWarnings({"rawtypes", "unchecked"})
public class HadoopApprovals
{
  public static void verifyMapping(SmartMapper mapper, Object key, Object input) throws Exception
  {
    MapDriver mapDriver = new MapDriver();
    mapDriver.setMapper(mapper);
    Object writableKey = WritableUtils.createWritable(key, mapper.getKeyInType());
    Object writableValue = WritableUtils.createWritable(input, mapper.getValueInType());
    mapDriver.withInput(writableKey, writableValue);
    List results = mapDriver.run();
    Collections.sort(results, PairComparer.INSTANCE);
    String header = String.format("[%s]\r\n\r\n -> maps via %s to -> \r\n", input, mapper.getClass()
        .getSimpleName());
    Approvals.verifyAll(header, results, Echo.INSTANCE);
  }
  public static <KeyIn, ValueIn, KeyOut, ValueOut> void verifyReducer(
      SmartReducer<KeyIn, ValueIn, KeyOut, ValueOut> reducer, Object key, Object... values) throws Exception
  {
    List<ValueIn> list = new ArrayList<ValueIn>();
    for (Object value : values)
    {
      list.add(WritableUtils.createWritable(value, reducer.getValueInType()));
    }
    ReduceDriver reduceDriver = new ReduceDriver<Text, LongWritable, Text, LongWritable>();
    reduceDriver.withInput(WritableUtils.createWritable(key, reducer.getKeyInType()), list);
    reduceDriver.setReducer(reducer);
    List results = reduceDriver.run();
    Collections.sort(results, PairComparer.INSTANCE);
    String header = String.format("(%s, %s)\r\n\r\n -> reduces via %s to -> \r\n", key, list, reducer.getClass()
        .getSimpleName());
    Approvals.verifyAll(header, results, Echo.INSTANCE);
  }
  public static void verifyMapReduce(SmartMapper mapper, SmartReducer reducer, Object key, Object input)
      throws Exception
  {
    MapDriver mapDriver = new MapDriver();
    mapDriver.setMapper(mapper);
    MapReduceDriver mapReduceDriver = new MapReduceDriver();
    mapReduceDriver.setMapper(mapper);
    Object writableKey = WritableUtils.createWritable(key, mapper.getKeyInType());
    Object writableValue = WritableUtils.createWritable(input, mapper.getValueInType());
    mapDriver.withInput(writableKey, writableValue);
    List results = mapDriver.run();
    Collections.sort(results, PairComparer.INSTANCE);
    mapReduceDriver = new MapReduceDriver<LongWritable, Text, Text, LongWritable, Text, LongWritable>();
    writableKey = WritableUtils.createWritable(key, mapper.getKeyInType());
    writableValue = WritableUtils.createWritable(input, mapper.getValueInType());
    mapReduceDriver.withInput(writableKey, writableValue);
    mapReduceDriver.setMapper(mapper);
    mapReduceDriver.setReducer(reducer);
    List finalResults = mapReduceDriver.run();
    String text = String.format(
        "[%s]\r\n\r\n -> maps via %s to -> \r\n\r\n%s\r\n\r\n -> reduces via %s to -> \r\n\r\n%s", input, mapper
            .getClass().getSimpleName(), ArrayUtils.toString(results, Echo.INSTANCE), reducer.getClass()
            .getSimpleName(), ArrayUtils.toString(finalResults, Echo.INSTANCE));
    Approvals.verify(text);
  }
}
