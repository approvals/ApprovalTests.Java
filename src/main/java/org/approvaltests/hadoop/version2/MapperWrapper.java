package org.approvaltests.hadoop.version2;

import java.io.IOException;

import org.apache.hadoop.mapred.JobConf;
import org.apache.hadoop.mapred.Mapper;
import org.apache.hadoop.mapred.OutputCollector;
import org.apache.hadoop.mapred.Reporter;

public class MapperWrapper<KeyIn, ValueIn, KeyOut, ValueOut>
    implements
      SmartMapper<KeyIn, ValueIn, KeyOut, ValueOut>
{
  private final Class<KeyIn>                             keyIn;
  private final Class<ValueIn>                           valueIn;
  private final Class<KeyOut>                            keyOut;
  private final Class<ValueOut>                          valueOut;
  private final Mapper<KeyIn, ValueIn, KeyOut, ValueOut> mapper;
  public MapperWrapper(Mapper<KeyIn, ValueIn, KeyOut, ValueOut> mapper, Class<KeyIn> keyIn,
      Class<ValueIn> valueIn, Class<KeyOut> keyOut, Class<ValueOut> valueOut)
  {
    this.mapper = mapper;
    this.keyIn = keyIn;
    this.valueIn = valueIn;
    this.keyOut = keyOut;
    this.valueOut = valueOut;
  }
  @Override
  public Class<KeyIn> getKeyInType()
  {
    return keyIn;
  }
  @Override
  public Class<ValueIn> getValueInType()
  {
    return valueIn;
  }
  @Override
  public Class<KeyOut> getKeyOutType()
  {
    return keyOut;
  }
  @Override
  public Class<ValueOut> getValueOutType()
  {
    return valueOut;
  }
  @Override
  public void map(KeyIn arg0, ValueIn arg1, OutputCollector<KeyOut, ValueOut> arg2, Reporter arg3)
      throws IOException
  {
    mapper.map(arg0, arg1, arg2, arg3);
  }
  @Override
  public void configure(JobConf arg0)
  {
    mapper.configure(arg0);
  }
  @Override
  public void close() throws IOException
  {
    mapper.close();
  }
}
