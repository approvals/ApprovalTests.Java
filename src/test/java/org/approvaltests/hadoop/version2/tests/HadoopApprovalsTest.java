package org.approvaltests.hadoop.version2.tests;

import java.io.IOException;
import java.util.Iterator;

import junit.framework.TestCase;

import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapred.MapReduceBase;
import org.apache.hadoop.mapred.Mapper;
import org.apache.hadoop.mapred.OutputCollector;
import org.apache.hadoop.mapred.Reducer;
import org.apache.hadoop.mapred.Reporter;
import org.approvaltests.hadoop.version2.HadoopApprovals;
import org.approvaltests.hadoop.version2.MapperWrapper;
import org.approvaltests.hadoop.version2.ReducerWrapper;
import org.approvaltests.hadoop.version2.SmartMapper;
import org.approvaltests.hadoop.version2.SmartReducer;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.UseReporter;

import com.spun.util.ArrayUtils;

@UseReporter({DiffReporter.class})
public class HadoopApprovalsTest extends TestCase
{
  public void testMap() throws Exception
  {
    SmartMapper mapper = new MapperWrapper(new WordCountMap(), LongWritable.class, Text.class, Text.class,
        LongWritable.class);
    HadoopApprovals.verifyMapping(mapper, 0, "llew gen  llew");
  }
  public void testReduce() throws Exception
  {
    SmartReducer reducer = new ReducerWrapper(new WordCountReduce(), Text.class, LongWritable.class, Text.class,
        LongWritable.class);
    HadoopApprovals.verifyReducer(reducer, "life", 7, 35);
  }
  public void testMapReducer() throws Exception
  {
    SmartMapper mapper = new MapperWrapper(new WordCountMap(), LongWritable.class, Text.class, Text.class,
        LongWritable.class);
    SmartReducer reducer = new ReducerWrapper(new WordCountReduce(), Text.class, LongWritable.class, Text.class,
        LongWritable.class);
    HadoopApprovals.verifyMapReduce(mapper, reducer, 0, "one fish two fish red fish blue fish");
  }
  public static class WordCountReduce extends MapReduceBase
      implements
        Reducer<Text, LongWritable, Text, LongWritable>
  {
    public void reduce(Text token, Iterator<LongWritable> counts, OutputCollector<Text, LongWritable> context,
        Reporter arg3) throws IOException
    {
      long n = 0;
      for (LongWritable count : ArrayUtils.asIterable(counts))
      {
        n += count.get();
      }
      context.collect(token, new LongWritable(n));
    }
  }
  private static class WordCountMap extends MapReduceBase
      implements
        Mapper<LongWritable, Text, Text, LongWritable>
  {
    public void map(LongWritable _, Text text, OutputCollector<Text, LongWritable> context, Reporter reporter)
        throws IOException
    {
      for (String token : text.toString().split("\\s+"))
      {
        context.collect(new Text(token), new LongWritable(1));
      }
    }
  }
}
