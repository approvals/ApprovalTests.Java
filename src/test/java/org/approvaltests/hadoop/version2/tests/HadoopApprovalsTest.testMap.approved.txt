[llew gen  llew]

 -> maps via MapperWrapper to -> 



(gen, 1)
(llew, 1)
(llew, 1)
